File: java-adaptor/src/main/java/org/judal/storage/java/TableOperation.java
package org.judal.storage.java;

import static org.judal.storage.query.SortDirection.ASC;
import static org.judal.storage.query.SortDirection.DESC;
import static org.judal.storage.query.SortDirection.same;

import javax.jdo.FetchGroup;
import javax.jdo.JDOException;
import javax.jdo.JDOUserException;

import org.judal.storage.table.AbstractTableOperation;
import org.judal.storage.table.Record;
import org.judal.storage.table.RecordSet;
import org.judal.storage.table.TableDataSource;

public class TableOperation<R extends Record> extends AbstractTableOperation<R> {

	public TableOperation() {
	}

	public TableOperation(R record) {
		 super(record);
	 }
	
	public TableOperation(TableDataSource dataSource) {
		super(dataSource);
	}

	public TableOperation(TableDataSource dataSource, R record) {
		super(dataSource, record);
	}
	
	public RecordSet<R> fetch(FetchGroup fetchGroup, String columnName, Object valueSearched) throws JDOException {
		return getTable().fetch(fetchGroup, columnName, valueSearched);
	}

	public RecordSet<R> fetchAsc(FetchGroup fetchGroup, String columnName, Object valueSearched, String sortByColumn) throws JDOException {
		RecordSet<R> retval = getTable().fetch(fetchGroup, columnName, valueSearched);
		retval.sort(sortByColumn);
		return retval;
	}

	public RecordSet<R> fetchDesc(FetchGroup fetchGroup, String columnName, Object valueSearched, String sortByColumn) throws JDOException {
		RecordSet<R> retval = getTable().fetch(fetchGroup, columnName, valueSearched);
		retval.sortDesc(sortByColumn);
		return retval;
	}

	@Override
	public R fetchFirst(FetchGroup fetchGroup, String columnName, Object valueSearched, String... sortBy)
		throws JDOException {
		RecordSet<R> rst;
		if (sortBy==null || sortBy.length==0)
			rst = fetch(fetchGroup, columnName, valueSearched);
		else if (sortBy.length==1)
			rst = fetchAsc(fetchGroup, columnName, valueSearched, sortBy[0]);
		else
			if (same(ASC,sortBy[1]))
				rst = fetchAsc(fetchGroup, columnName, valueSearched, sortBy[0]);
			else if (same(DESC,sortBy[1]))
				rst = fetchDesc(fetchGroup, columnName, valueSearched, sortBy[0]);
			else
				throw new JDOUserException("Unrecognized sort direction " + sortBy[1]);
		setRecord(rst.size()>0 ? rst.get(0) : null);
		return getRecord();
	}

}

File: scala-adaptor/src/main/scala/org/judal/storage/scala/TableOperation.scala
package org.judal.storage.scala

import javax.jdo.FetchGroup
import javax.jdo.JDOUserException

import org.judal.storage.EngineFactory
import org.judal.storage.table.AbstractTableOperation
import org.judal.storage.table.Record
import org.judal.storage.table.RecordSet
import org.judal.storage.table.TableDataSource

import org.judal.storage.query.SortDirection.{ASC,DESC,same}

import scala.collection.JavaConverters._

class TableOperation[R >: Null <: Record](dataSource: TableDataSource , record: R) extends AbstractTableOperation[R](dataSource, record) {

	def this(record: R) = this(EngineFactory.getDefaultTableDataSource, record)

	def this(dataSource: TableDataSource) = this(dataSource, null)

	def fetch(fetchGroup: FetchGroup, columnName: String , valueSearched: AnyRef)  : Iterable[R] = {
		getTable.fetch(fetchGroup, columnName, valueSearched).asScala
	}
	
	def fetchAsc(fetchGroup: FetchGroup, columnName: String , valueSearched: AnyRef, sortByColumn: String) : Iterable[R]  = {
		val retval : RecordSet[R] = getTable.fetch(fetchGroup, columnName, valueSearched)
		retval.sort(sortByColumn)
		retval.asScala
	}
	
	def fetchDesc(fetchGroup: FetchGroup, columnName: String , valueSearched: AnyRef, sortByColumn: String) : Iterable[R]  = {
		val retval : RecordSet[R] = getTable.fetch(fetchGroup, columnName, valueSearched)
		retval.sortDesc(sortByColumn)
		retval.asScala
	}

	override def fetchFirst(fetchGroup: FetchGroup , columnName: String , valueSearched: Any, sortBy: String*) : R = {
		var rst: RecordSet[R] = getTable.fetch(fetchGroup, columnName, valueSearched)
		if (rst.size()>0) {
		  if (sortBy!=null && sortBy.length==1)
		    rst.sort(sortBy(0))
		  else if (sortBy!=null && sortBy.length>1) {
			  if (same(ASC,sortBy(1)))
		      rst.sort(sortBy(0))
			  else if (same(DESC,sortBy(1)))
			    rst.sortDesc(sortBy(0))
			  else
				  throw new JDOUserException("Unrecognized sort direction " + sortBy(1))
		  }
		  val retval : R = rst.get(0)
		  setRecord(retval)
		} else {
		  setRecord(null)
		}
		getRecord()
	}
}